File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-4.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int l = 0;
        
        int ws = 0;
        int ms = INT_MAX;
        bool d = false; 
        for (int r = 0; r < nums.size(); r++) {
            ws += nums[r];
            while (ws >= target) {
                if(!d) d = true;
                ms = min(ms, r-l+1);
                ws -= nums[l++];
            }
        }

        return d ? ms : 0;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-4.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        neg_max = 0
        pos_max =  nums[0]
        ncmax = nums[0]

        for num in nums[1:]:
            c = (num * neg_max, num * pos_max, num)
            neg_max = min(c)
            pos_max = max(c)
            ncmax = max(pos_max, ncmax)
        return ncmax